pkg/logql/series_extractor.go

```diff
@@ -0,0 +1,99 @@
+package logql
+
+import (
+	"github.com/grafana/loki/pkg/iter"
+	"github.com/grafana/loki/pkg/logproto"
+)
+
+var (
+	extractBytes = bytesSampleExtractor{}
+	extractCount = countSampleExtractor{}
+)
+
+// SeriesIterator is an iterator that iterate over a stream of logs and returns sample.
+type SeriesIterator interface {
+	Close() error
+	Next() bool
+	Peek() (Sample, bool)
+}
+
+// Sample is a series sample
+type Sample struct {
+	Labels        string
+	Value         float64
+	TimestampNano int64
+}
+
+type seriesIterator struct {
+	iter    iter.PeekingEntryIterator
+	sampler SampleExtractor
+
+	updated bool
+	cur     Sample
+}
+
+func newSeriesIterator(it iter.EntryIterator, sampler SampleExtractor) SeriesIterator {
+	return &seriesIterator{
+		iter:    iter.NewPeekingIterator(it),
+		sampler: sampler,
+	}
+}
+
+func (e *seriesIterator) Close() error {
+	return e.iter.Close()
+}
+
+func (e *seriesIterator) Next() bool {
```

Comment:
I believe this should also consume from the underlying iter until the extractor is successful or the iterator is exhausted. This would ensure the `*seriesIterator` behaves correctly regardless of if it was `Peek`'ed or not.

Reply:
Not fully sure what you mean but it does consume it via the PeekingIterator.

Reply:
If the `seriesIterator` embeds a `SampleExtractor`, why does the `Next()` method not use it -- that seems incorrect to me.

Reply:
The interface is not the same as an `EntryIterator` its this:

```
// SeriesIterator is an iterator that iterate over a stream of logs and returns sample.
type SeriesIterator interface {
	Close() error
	Next() bool
	Peek() (Sample, bool)
}
```

The only way to get a sample if to use Peek. Next move forward only.

Reply:
I hear you though about the fact that Next could probably advance faster by trying the conversion. But since the only way to get a sample is Peek, I prefered to keep it simple and only have the complex logic in peek.

If the interface was different for example this:
```
// SeriesIterator is an iterator that iterate over a stream of logs and returns sample.
type SeriesIterator interface {
         Sample() Sample
	Close() error
	Next() bool
	Peek() (Sample, bool)
}
```

Then yeah something would be wrong cause next would not populate the current sample.

Reply:
Ah, thanks for explaining -- I conflated it with our `EntryIterator` interface.